envi/__init__.py

```diff
@@ -1349,7 +1381,103 @@ def getCurrentArch():
 
     return ret
 
+def getRealArchName(name):
+    for rname, aliases in arch_aliases.items():
+        if name in aliases:
+            return rname
+
+    return name
+
+def getArchNames():
+    """
+    Return a dict of arch_names which are currently implemented
+
+    envi.arch_names may hold future architectures which have not been merged in yet
+    """
+    modules = getArchModules(new=False)
+    return {arch: name for (arch, name) in arch_names.items() if arch>>16 < len(modules) and modules[arch>>16]}
+    
+
 def getArchModule(name=None):
+    """
+    Return an Envi architecture module instance for the following
+    architecture name.
+
+    If name is None, uses the "current" architecture.
+    """
+    if name is None:
+        name = getCurrentArch()
+
+    rname = getRealArchName(name)
+    archnum = arch_by_name.get(rname)
+
+    if archnum not in arch_module_path:
+        raise ArchNotImplemented(name)
+
+    # retrieve path and class info.  envi/archs/<aname>/__init__.py with amodname()
+    aname, amodname = arch_module_path.get(archnum)
+
+    modpath = os.sep.join(['envi', 'archs', aname, '__init__.py'])
+    if not os.path.exists(modpath):
+        raise ArchNotImplemented(name)
+
+    # get the module spec
+    spec = imputil.spec_from_file_location(rname, os.sep.join(['envi','archs', aname, '__init__.py']))
+    if not spec:
+        raise ArchNotImplemented(name)
+
+    # create an unintialized module from the spec
+    module = imputil.module_from_spec(spec)
+    if not module:
+        raise ArchNotImplemented(name)
+
+    # insert the module into sys.modules:
+    sys.modules[rname] = module
+
+    # initialize the module (actually "importing" it)
+    spec.loader.exec_module(module)
+
+    # instantiate the ArchitectureModule
+    cls = getattr(module, amodname)
+    archmod = cls()
+    
+    return archmod
+
+def getArchModules(default=ARCH_DEFAULT, new=True):
+    '''
+    Retrieve a default array of arch modules ( where index 0 is
+    also the "named" or "default" arch module.
+    '''
+    global shared_modules
```

Comment:
I get the need for caching, but I think getArchModules is rare enough to where we don't need to complicate this with a global cache variable to maintain some state.

Reply:
None.  Will fix. Thanks

Reply:
i did this to assuage my guilt over accessing it *every* time we create a context menu.

Reply:
this is now the way we know if an architecture has been implemented.
although, perhaps this becomes a conversation of the dict of dicts...

Reply:
I think this becomes a conversation of the dict of dicts and yea, we can wait for that.